File: src/controllers/postComment.controllers.js
import * as postCommentService from "../services/postComment.service.js"

export const createNewPost = async (req, res) => {
  try {
    const { media_data_list, type, description = "" } = req.body

    const { client_username } = req.auth

    const resp = await postCommentService.createNewPost({
      client_username,
      media_data_list,
      type,
      description,
    })

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const reactToPost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { reaction } = req.body

    const { client_username } = req.auth

    const resp = await postCommentService.reactToPost({
      client_username,
      post_id,
      reaction,
    })

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const commentOnPost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { comment_text, attachment_data } = req.body

    const { client_username } = req.auth

    const resp = await postCommentService.commentOnPost({
      client_username,
      post_id,
      comment_text,
      attachment_data,
    })

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const reactToComment = async (req, res) => {
  try {
    const { comment_id } = req.params
    const { reaction } = req.body

    const { client_username } = req.auth

    const resp = await postCommentService.reactToComment({
      client_username,
      comment_id,
      reaction,
    })

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const commentOnComment = async (req, res) => {
  try {
    const { comment_id } = req.params
    const { comment_text, attachment_data } = req.body

    const { client_username } = req.auth

    const resp = await postCommentService.commentOnComment({
      client_username,
      comment_id,
      comment_text,
      attachment_data,
    })

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const createRepost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.createRepost(post_id, client_username)

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const savePost = async (req, res) => {
  try {
    const { post_id } = req.params

    const { client_username } = req.auth

    const resp = await postCommentService.savePost(post_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const unsavePost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.unsavePost(post_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

/* The GETs */

export const getPost = async (req, res) => {
  try {
    const { post_id } = req.params

    const { client_username } = req.auth

    const resp = await postCommentService.getPost(post_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getCommentsOnPost = async (req, res) => {
  try {
    const { post_id } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getCommentsOnPost({
      post_id,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getComment = async (req, res) => {
  try {
    const { comment_id } = req.params

    const { client_username } = req.auth

    const resp = await postCommentService.getComment(comment_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getReactorsToPost = async (req, res) => {
  try {
    const { post_id } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getReactorsToPost({
      post_id,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getReactorsWithReactionToPost = async (req, res) => {
  try {
    const { post_id, reaction } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getReactorsWithReactionToPost({
      post_id,
      reaction,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getCommentsOnComment = async (req, res) => {
  try {
    const { comment_id } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getCommentsOnComment({
      comment_id,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getReactorsToComment = async (req, res) => {
  try {
    const { comment_id } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getReactorsToComment({
      comment_id,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const getReactorsWithReactionToComment = async (req, res) => {
  try {
    const { comment_id, reaction } = req.params

    const { limit = 20, offset = 0 } = req.query

    const { client_username } = req.auth

    const resp = await postCommentService.getReactorsWithReactionToComment({
      comment_id,
      reaction,
      client_username,
      limit,
      offset,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

/* DELETEs */

export const deletePost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.deletePost(post_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const removeReactionToPost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.removeReactionToPost(
      post_id,
      client_username
    )

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const removeCommentOnPost = async (req, res) => {
  try {
    const { post_id, comment_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.removeCommentOnPost({
      post_id,
      comment_id,
      client_username,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const removeCommentOnComment = async (req, res) => {
  try {
    const { parent_comment_id, comment_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.removeCommentOnComment({
      parent_comment_id,
      comment_id,
      client_username,
    })

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const removeReactionToComment = async (req, res) => {
  try {
    const { comment_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.removeReactionToComment(
      comment_id,
      client_username
    )

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const deleteRepost = async (req, res) => {
  try {
    const { post_id } = req.params
    const { client_username } = req.auth

    const resp = await postCommentService.deletePost(post_id, client_username)

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

File: src/controllers/postCommentControllers/postCommentControllers.go
package postCommentControllers

import (
	"i9lyfe/src/appTypes"
	"i9lyfe/src/helpers"
	"i9lyfe/src/services/postCommentService"

	"github.com/gofiber/fiber/v3"
)

func AuthorizePostUpload(c fiber.Ctx) error {
	ctx := c.Context()

	var body authorizePostUploadBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(); err != nil {
		return err
	}

	respData, err := postCommentService.AuthorizePostMediaUpload(ctx, body.PostType, body.MediaMIME, len(body.MediaSizes))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func AuthorizeCommentUpload(c fiber.Ctx) error {
	ctx := c.Context()

	var body authorizeCommentUploadBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(); err != nil {
		return err
	}

	respData, err := postCommentService.AuthorizeCommAttUpload(ctx, body.AttachmentMIME)
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func CreateNewPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var body createNewPostBody

	var err error

	err = c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(ctx); err != nil {
		return err
	}

	respData, err := postCommentService.CreateNewPost(ctx, clientUser.Username, body.MediaCloudNames, body.Type, body.Description, body.At)
	if err != nil {
		return err
	}

	return c.Status(201).MsgPack(respData)
}

func GetPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.GetPost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func DeletePost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.DeletePost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func ReactToPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var err error

	var body reactToPostBody

	err = c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(); err != nil {
		return err
	}

	respData, err := postCommentService.ReactToPost(ctx, clientUser.Username, c.Params("postId"), body.Emoji, body.At)
	if err != nil {
		return err
	}

	return c.Status(201).MsgPack(respData)
}

func GetReactorsToPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var query struct {
		Limit  int64
		Cursor float64
	}

	if err := c.Bind().Query(&query); err != nil {
		return err
	}

	respData, err := postCommentService.GetReactorsToPost(ctx, clientUser.Username, c.Params("postId"), helpers.CoalesceInt(query.Limit, 20), query.Cursor)
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

/* func GetReactorsWithReactionToPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	reaction, err := url.PathUnescape(c.Params("reaction"))
	if err != nil {
		return fiber.NewError(fiber.StatusBadRequest, fmt.Sprint(err))
	}

	respData, err := postCommentService.GetReactorsWithReactionToPost(ctx, clientUser.Username, c.Params("postId"), reaction, c.QueryInt("limit", 20), int64(c.QueryInt("offset")))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
} */

func RemoveReactionToPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.RemoveReactionToPost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func CommentOnPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var err error

	var body commentOnPostBody

	err = c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(ctx); err != nil {
		return err
	}

	respData, err := postCommentService.CommentOnPost(ctx, clientUser.Username, c.Params("postId"), body.CommentText, body.AttachmentCloudName, body.At)
	if err != nil {
		return err
	}

	return c.Status(201).MsgPack(respData)
}

func GetCommentsOnPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var query struct {
		Limit  int64
		Cursor float64
	}

	if err := c.Bind().Query(&query); err != nil {
		return err
	}

	respData, err := postCommentService.GetCommentsOnPost(ctx, clientUser.Username, c.Params("postId"), helpers.CoalesceInt(query.Limit, 20), query.Cursor)
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func GetComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.GetComment(ctx, clientUser.Username, c.Params("commentId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func RemoveCommentOnPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.RemoveCommentOnPost(ctx, clientUser.Username, c.Params("postId"), c.Params("commentId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func ReactToComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var err error

	var body reactToCommentBody

	err = c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(); err != nil {
		return err
	}

	respData, err := postCommentService.ReactToComment(ctx, clientUser.Username, c.Params("commentId"), body.Emoji, body.At)
	if err != nil {
		return err
	}

	return c.Status(201).MsgPack(respData)
}

func GetReactorsToComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var query struct {
		Limit  int64
		Cursor float64
	}

	if err := c.Bind().Query(&query); err != nil {
		return err
	}

	respData, err := postCommentService.GetReactorsToComment(ctx, clientUser.Username, c.Params("commentId"), helpers.CoalesceInt(query.Limit, 20), query.Cursor)
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

/* func GetReactorsWithReactionToComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	reaction, err := url.PathUnescape(c.Params("reaction"))
	if err != nil {
		return fiber.NewError(fiber.StatusBadRequest, fmt.Sprint(err))
	}

	respData, err := postCommentService.GetReactorsWithReactionToComment(ctx, clientUser.Username, c.Params("commentId"), reaction, c.QueryInt("limit", 20), int64(c.QueryInt("offset")))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
} */

func RemoveReactionToComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.RemoveReactionToComment(ctx, clientUser.Username, c.Params("commentId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func CommentOnComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var err error

	var body commentOnCommentBody

	err = c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err = body.Validate(ctx); err != nil {
		return err
	}

	respData, err := postCommentService.CommentOnComment(ctx, clientUser.Username, c.Params("commentId"), body.CommentText, body.AttachmentCloudName, body.At)
	if err != nil {
		return err
	}

	return c.Status(201).MsgPack(respData)
}

func GetCommentsOnComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	var query struct {
		Limit  int64
		Cursor float64
	}

	if err := c.Bind().Query(&query); err != nil {
		return err
	}

	respData, err := postCommentService.GetCommentsOnComment(ctx, clientUser.Username, c.Params("commentId"), helpers.CoalesceInt(query.Limit, 20), query.Cursor)
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func RemoveCommentOnComment(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.RemoveCommentOnComment(ctx, clientUser.Username, c.Params("parentCommentId"), c.Params("childCommentId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func RepostPost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.RepostPost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func SavePost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.SavePost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}

func UnsavePost(c fiber.Ctx) error {
	ctx := c.Context()

	clientUser := c.Locals("user").(appTypes.ClientUser)

	respData, err := postCommentService.UnsavePost(ctx, clientUser.Username, c.Params("postId"))
	if err != nil {
		return err
	}

	return c.MsgPack(respData)
}
